build: fix preview-docs command in Docker

Juneezee opened the pull request:
## What/Why/How?

As described in #272, running `preview-docs` in a `redocly/openapi-cli` Docker container will cause the container to crash instantly. This is not an issue with the Docker image. The problem lies in the Webpack build process that is causing the following `TypeError`. 

Running the following command `npm run bundle && node dist/bundle.js preview-docs resources/pets.yaml` will result in the following error as well.

```
TypeError: s is not a constructor
    at Object.t.startWsServer (/bin/openapi-cli:49:129227)
    at Object.<anonymous> (/bin/openapi-cli:49:21717)
    at Generator.next (<anonymous>)
    at o (/bin/openapi-cli:49:19069)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
```

---
Currently Webpack will load `simple-websocket` as an empty module. This is introduced in PR #177, which was merged into the `master` branch before the Dockerfile was first created in commit ceaa205497a6cee6a02876a75b8ddcd4e37aa23c. 
https://github.com/Redocly/openapi-cli/blob/360c865d23668c46cad9ee63677fce54925ce238/webpack.config.ts#L20-L26

---
However, `preview-docs` requires the websocket server to be started. If `simple-websocket` is loaded as an empty module, then `SocketServer` will have the value `{}` and hence `new SocketServer` will throw a `TypeError`.

https://github.com/Redocly/openapi-cli/blob/38139e51393d7589217362986b36fb48771ab703/packages/cli/src/commands/preview-docs/preview-server/server.ts#L5

https://github.com/Redocly/openapi-cli/blob/38139e51393d7589217362986b36fb48771ab703/packages/cli/src/commands/preview-docs/preview-server/server.ts#L65-L66

---

To fix this, we would need to configure Webpack to load `simple-websocket`. The steps are:

1. Use `babel-loader` to load the `simple-websocket` module. We will need to use the `@babel/plugin-proposal-class-properties` plugin as well, otherwise the following error will occur.
```
ERROR in ./node_modules/simple-websocket/server.js 45:19
Module parse failed: Unexpected token (45:19)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
|   }
| 
>   _handleListening = () => {
|     this.emit('listening')
|   }
 @ ./packages/cli/src/commands/preview-docs/preview-server/server.ts 6:21-58
 @ ./packages/cli/src/commands/preview-docs/preview-server/preview-server.ts
 @ ./packages/cli/src/commands/preview-docs/index.ts
 @ ./packages/cli/src/index.ts
```

2. Configure Webpack to use named instead of number for Module ID. `require.resolve` returns a number instead of a string by Webpack for optimization purposes (see [Webpack 4 require.resolve](https://v4.webpack.js.org/api/module-methods/#requireresolve) and [Webpack 4 Module IDs](https://v4.webpack.js.org/configuration/optimization/#optimizationmoduleids)). This is required otherwise line 107 will fail because `filePath` is a number, and `path.extname` expects a string.
https://github.com/Redocly/openapi-cli/blob/38139e51393d7589217362986b36fb48771ab703/packages/cli/src/commands/preview-docs/preview-server/preview-server.ts#L99-L107

3. Copy the `packages/cli/src/commands/preview-docs/preview-server/default.hbs` file into `dist` folder, and include it into the Docker image too.

## Reference

This pull request fixes Issue #272.

## Testing

## Screenshots (optional)

### Before
![before](https://user-images.githubusercontent.com/20135478/132394577-fb910ed1-44a8-4d16-8121-19a2d088fffa.png)

### After
![after](https://user-images.githubusercontent.com/20135478/132394704-a75125aa-61fd-48f4-8cdf-3e030fbed9d1.png)

## Check yourself

- [x] Code is linted
- [x] Tested with redoc/reference-docs/workflows
- [x] All new/updated code is covered with tests

## Security

- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

RomanHotsiy commented:
Great job on investigation here. I'm wandering if it's not easier to just copy unbundled sources into the docker container vs introducing more build steps.

@Juneezee what do you think? Would you be open to help with this?

Juneezee commented:
@RomanHotsiy Thank you for taking your time to review this PR :heart: !

> Great job on investigation here. I'm wandering if it's not easier to just copy unbundled sources into the docker container vs introducing more build steps.
> 
> @Juneezee what do you think? Would you be open to help with this?

Yes, the easiest way to fix #272 is actually building the Docker image using `npm install -g @redocly/openapi-cli`, and we don't need to add those `babel` into `package.json`. However, the final Docker image will be around 10 ~ 20 MB larger than the image built with `npm run webpack-bundle`.

If this is acceptable for the Redoc team, then I am happy to modify this PR to use the easy method mentioned above :smiley: .

RomanHotsiy commented:
> building the Docker image using `npm install -g @redocly/openapi-cli`

That will require package to be published on NPM first. I've seen npm may be lagging behind because of the cache sometimes. Please, keep it in mind.

> If this is acceptable for the Redoc team, then I am happy to modify this PR to use the easy method mentioned above 😃 .

This is acceptable. Please, modify the PR 🙏

Juneezee commented:
@RomanHotsiy Please take a look again. Thank you.

Juneezee commented:
@RomanHotsiy I'm sorry for the back and forth changes. I have just figured out a better way to build the Docker image without relying on `npmjs.com`. Now the Docker image can be built without having to wait package being published first to `npmjs.com`.

Please take another look again. Thanks.

RomanHotsiy commented:
Looks much better now, the only thing left is to split RUN into multiple commands to get proper caching, see this: https://deploy-docs.aptible.com/docs/dockerfile-caching#npm

Juneezee commented:
I have split the `RUN` commands and added some comments. I did some quick googling for Docker image build cache and found some example best practices demonstrated in

1. https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#run
2. https://stackoverflow.com/a/50097098/7902371

I am not experienced enough with Docker build cache :cry: so I still couldn't find the correct boundary of whether to join or split multiple `RUN` commands

RomanHotsiy commented:
Thanks! I will update it slightly later this week.

RomanHotsiy commented:
Hey @Juneezee, take a look at my changes here b75b22c

If they look okay, let me know and I'll merge it.

Juneezee commented:
@RomanHotsiy Looks good to me. Thanks
